### PR TITLE
Moretables review proposals

### LINKED ISSUE #133: Add adapter property for maximum number of tables
The table limit of 1000 introduced in version 9.0.0 was arbitrary and meant to protect against unexpected refresh duration or internal message size overflow.

However, if the user knows what they are doing, an override is needed. We agreed on making the limit configurable instead of adding a "ignore the limit" parameter.

### PR DESCRIPTION
**Pull request by ckunki**
Fixes #133

### CODE REVIEW

File: src/main/java/com/exasol/adapter/jdbc/BaseTableMetadataReader.java [hunk 1]
```diff
@@ -72,7 +70,7 @@ private List<TableMetadata> extractTableMetadata(final ResultSet remoteTables, f
         do {
             final Optional<TableMetadata> tableMetadata = getTableMetadata(remoteTables, filteredTables);
             tableMetadata.ifPresent(mappedTables::add);
-            validateMappedTablesListSize(mappedTables);
+            TableCountLimit.from(this.properties).validateNumberOfTables(mappedTables.size());
```
Comment (exaSR) [inline src/main/java/com/exasol/adapter/jdbc/BaseTableMetadataReader.java:73]: Move from() part outside the loop, assuming properties do not change.

File: src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java [hunk 2]
```diff
@@ -184,7 +184,10 @@ public void validateProperties() throws PropertyValidationException {
         validateSchemaNameProperty();
         validateDebugOutputAddress();
         validateExceptionHandling();
-        validateDataTypeDetectionStrategy();
+        PropertyValidator.chain() //
+                .add(DataTypeDetection.getValidator()) //
+                .add(TableCountLimit.getValidator()) //
```
Comment (exaSR) [inline src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java:189]: Not sure, but would it make sense to put the chain generation into the constructor? That way each Adapter/Dialect could simply add its own validators there, and the validate method could be `final`

File: src/main/java/com/exasol/adapter/jdbc/JDBCAdapter.java [hunk 3]
```diff
@@ -121,78 +121,39 @@ private SchemaMetadata getRemoteMetadata(final ExaMetadata metadata, final Refre
...
     }
 
     @Override
     public SetPropertiesResponse setProperties(final ExaMetadata metadata, final SetPropertiesRequest request)
             throws AdapterException {
         final Map<String, String> requestRawProperties = request.getProperties();
+
+        if (!PropertyValidator.requiresRefreshOfVirtualSchema(requestRawProperties)) {
+            return SetPropertiesResponse.builder().schemaMetadata(null).build();
+        }
```
Comment (exaSR) [inline src/main/java/com/exasol/adapter/jdbc/JDBCAdapter.java:140]: Does this not skip validation again?

**Review by exaSR — CHANGES_REQUESTED** [review]
I love this! Thanks for investing so much time to refactor the validation and making an actual class for the parameter including its behavior!

However, now that you have opened the can of worms... should not most of the new code be put into `virtual-schema-java`? :)

File: src/main/java/com/exasol/adapter/properties/PropertyValidator.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * @param properties
+     * @return {@code true} if one of the keys of parameter {properties} requires to refresh the virtual schema.
+     */
+    public static boolean requiresRefreshOfVirtualSchema(final Map<String, String> properties) {
+        return properties.containsKey(TableCountLimit.MAXTABLES_PROPERTY)
+                || AdapterProperties.isRefreshingVirtualSchemaRequired(properties);
+    }
```
Comment (exaSR) [inline src/main/java/com/exasol/adapter/properties/PropertyValidator.java:27]: ah, this is the one that disturbed me during my review... The core `PropertyValidator` class should not have to know about any specific properties...

File: src/main/java/com/exasol/adapter/jdbc/BaseTableMetadataReader.java (hunk 1 above)
Comment (ckunki) [reply src/main/java/com/exasol/adapter/jdbc/BaseTableMetadataReader.java:73]: 👍

File: src/main/java/com/exasol/adapter/jdbc/JDBCAdapter.java (hunk 3 above)
Comment (ckunki) [reply src/main/java/com/exasol/adapter/jdbc/JDBCAdapter.java:140]: You are right, I will change it again.

File: src/main/java/com/exasol/adapter/properties/PropertyValidator.java (hunk 4 above)
Comment (ckunki) [reply src/main/java/com/exasol/adapter/properties/PropertyValidator.java:27]: OK.

File: src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java (hunk 2 above)
Comment (ckunki) [reply src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java]: I agree, but was too lazy until now 😃 I will update the PR.

**Comment by ckunki** [comment]
> I love this! Thanks for investing so much time to refactor the validation and making an actual class for the parameter including its behavior!

Thanks - this is a motivating reward for me.

> However, now that you have opened the can of worms... should not most of the new code be put into `virtual-schema-java`? :)

Maybe. I am not sure, yet. But I would leave that as a separate ticket / PR, if you are OK with that.

File: src/main/java/com/exasol/adapter/properties/TableCountLimit.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public static PropertyValidator getValidator() {
+        return PropertyValidator.optional(MAXTABLES_PROPERTY, TableCountLimit::validatePropertyValue);
+    }
+
+    private static void validatePropertyValue(final String value) throws PropertyValidationException {
+        try {
+            final int parsed = Integer.parseUnsignedInt(value);
+            if (parsed == 0) {
+                throw new IllegalArgumentException();
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/properties/TableCountLimit.java:32]: Should we use a better message, e.g. "must be greater than 0"?

File: src/main/java/com/exasol/adapter/properties/TableCountLimit.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            }
+        } catch (final IllegalArgumentException exception) {
+            throw new PropertyValidationException(ExaError.messageBuilder("E-VSCJDBC-43") //
+                    .message("Invalid parameter value.") //
+                    .mitigation("The adapter property {{max_tables_property}}" //
+                            + " if present, must be a positive integer.") //
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/properties/TableCountLimit.java:38]:
```suggestion
                    .mitigation("The adapter property {{max_tables_property}}" //
                            + " if present, must be a positive integer greater than 0.") //
```

File: src/main/java/com/exasol/adapter/properties/TableCountLimit.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+     * @param maxNumberOfTables maximum number of tables to be accepted
+     */
+    public TableCountLimit(final int maxNumberOfTables) {
+        this.maxNumberOfTables = maxNumberOfTables;
+    }
+
+    /**
+     * Verify that the given number of mapped tables does not exceed the configured max table limit.
+     *
+     * @param numberOfTables actual size of mapped tables
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/properties/TableCountLimit.java:68]:
```suggestion
     * @param numberOfTables actual number of mapped tables
```

File: src/main/java/com/exasol/adapter/properties/TableCountLimit.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        if (numberOfTables > this.maxNumberOfTables) {
+            throw new RemoteMetadataReaderException(ExaError.messageBuilder("E-VSCJDBC-42")
+                    .message("The size of the list of the selected tables exceeds" //
+                            + " the configured allowed maximum of {{current_limit}}.")
+                    .mitigation(
+                            "Please use the {{table_filter_property}} property to define the list of tables you need" //
+                                    + " or increase the limit using the {{max_tables_property}} property.") //
+                    .parameter("current_limit", this.maxNumberOfTables) //
+                    .parameter("table_filter_property", TABLE_FILTER_PROPERTY) //
+                    .parameter("max_tables_property", MAXTABLES_PROPERTY) //
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/properties/TableCountLimit.java:81]: No need to use parameters for constants. Then the message is easier to find in the error catalog. Just use string concatenation.

File: src/main/java/com/exasol/adapter/properties/ValidatorChain.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+
+import com.exasol.adapter.AdapterProperties;
+
+/**
+ * Validate the specified properties by applying all property validators subsequently.
+ */
+public class ValidatorChain {
+
+    List<PropertyValidator> propertyValidators = new ArrayList<>();
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/properties/ValidatorChain.java:13]:
```suggestion
    private final List<PropertyValidator> propertyValidators = new ArrayList<>();
```

File: src/main/java/com/exasol/adapter/properties/ValidatorChain.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.exasol.adapter.AdapterProperties;
+
+/**
+ * Validate the specified properties by applying all property validators subsequently.
+ */
+public class ValidatorChain {
```
Comment (kaklakariada) [inline src/main/java/com/exasol/adapter/properties/ValidatorChain.java:11]:
```suggestion
public class ValidatorChain implements PropertyValidator {
```
Would be nice if possible

File: src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java [hunk 11]
```diff
@@ -103,15 +105,61 @@ void testWarnIfTableScanResultsAreEmpty() throws SQLException {
...
+        final List<String> noFilter = Collections.emptyList();
+        final RemoteMetadataReaderException exception = assertThrows(RemoteMetadataReaderException.class,
+                () -> metadataReader.mapTables(resultSetMock, noFilter));
+        assertAll( //
+                () -> assertThat(exception.getMessage(), containsString("E-VSCJDBC-42")), //
+                () -> assertThat(exception.getMessage(), containsString("1000")) //
+        );
+    }
+
+    private ResultSet resultSetWithUnlimitedSize() throws SQLException {
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java:121]: 👍

File: src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java [hunk 12]
```diff
@@ -103,15 +105,61 @@ void testWarnIfTableScanResultsAreEmpty() throws SQLException {
...
+        assertAll(() -> assertThat(exception.getMessage(), containsString("E-VSCJDBC-42")),
+                () -> assertThat(exception.getMessage(), containsString("2000")));
+    }
+
+    // verify that it does map 3000 tables when the parameter is set so
+    @Test
+    void testValidateMappedTablesListSizeWithProperty3000() throws SQLException {
+        final ResultSet resultSetMock = Mockito.mock(ResultSet.class);
+        // limit number of returned tables, so we can confirm it does map 3000 tables
+        final AtomicInteger tableNumber = new AtomicInteger(0);
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java:150]:
```suggestion
        final AtomicInteger tableCount= new AtomicInteger(0);
```
Also extract configuration of the mock to a method like `resultSetWithUnlimitedSize()`.
Then you can also add a test where the result set contains one table more than the limit.

File: src/test/java/com/exasol/adapter/properties/BooleanPropertyTest.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.Map;
+
+import org.junit.jupiter.api.Test;
+
+import com.exasol.adapter.AdapterProperties;
+
+class BooleanPropertyTest {
+    @Test
+    void testValidateBooleanProperty() {
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/properties/BooleanPropertyTest.java:16]: Also add a 'good' case test where validation succeeds

File: src/test/java/com/exasol/adapter/properties/DataTypeDetectionTest.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        final AdapterProperties properties = new AdapterProperties(Map.of());
+        final DataTypeDetection testee = DataTypeDetection.from(properties);
+        assertThat(testee.getStrategy(), equalTo(DataTypeDetection.DEFAULT_STRATEGY));
+        verifyValidator(properties);
+    }
+
+    @Test
+    void testValidationFailure() throws PropertyValidationException {
+        final PropertyValidator validator = DataTypeDetection.getValidator();
+        assertThrows(PropertyValidationException.class, () -> validator.validate(adapterProperties("invalid_value")));
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/properties/DataTypeDetectionTest.java:38]: Add an assertion for the error message

File: src/test/java/com/exasol/adapter/properties/DataTypeDetectionTest.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    private void verifyValidator(final AdapterProperties properties) {
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/properties/DataTypeDetectionTest.java:41]:
```suggestion
    private void verifyValidatedSuccessfully(final AdapterProperties properties) {
```
❓

File: src/test/java/com/exasol/adapter/properties/TableCountLimitTest.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        verify(testee, 1000);
+        verifyValidator(properties);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = { "-1", "0", "1.2", "2,3" })
+    void testValidatorFailure(final String value) {
+        final Map<String, String> properties = Map.of(TableCountLimit.MAXTABLES_PROPERTY, value);
+        final PropertyValidator validator = TableCountLimit.getValidator();
+        assertThrows(PropertyValidationException.class, () -> validator.validate(new AdapterProperties(properties)));
```
Comment (kaklakariada) [inline src/test/java/com/exasol/adapter/properties/TableCountLimitTest.java:44]: Add assertion for error message

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java [hunk 17]
```diff
@@ -359,20 +264,12 @@ protected void checkImportPropertyConsistency(final String importFromProperty, f
      *
      * @param castNumberToDecimalProperty property name
      * @throws PropertyValidationException if validation fails
+     *
+     * @deprecated Please use {@code com.exasol.adapter.CastNumberToDecimalProperty.validator(String)} instead.
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java:268]: 👍

File: src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java [hunk 18]
```diff
@@ -206,121 +206,26 @@ protected void validateSupportedPropertiesList() throws PropertyValidationExcept
...
     }
 
     /**
      * Validate the input of a boolean property.
      *
-     * @param property property name
+     * @param propertyName property name
      * @throws PropertyValidationException if validation fails
+     *
+     * @deprecated Please use {@code com.exasol.adapter.propertiesBooleanProperty.validator(String)} instead.
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java:224]: 👍

File: src/main/java/com/exasol/adapter/dialects/validators/ConnectionNameProperty.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * This class enables to validate the value of the connection name property.
+ */
+public class ConnectionNameProperty implements PropertyValidator {
+    /**
+     * @return new instance of {@link PropertyValidator} for validation of connection name property.
+     */
+    public static PropertyValidator validator() {
+        return PropertyValidator.ignoreEmpty(DEBUG_ADDRESS_PROPERTY, new DebugPortNumberProperty());
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/adapter/dialects/validators/ConnectionNameProperty.java:19]: Ups - thank you so much. The static method was not called so the problem was hidden.
I made the constructor private now.

File: src/test/java/com/exasol/adapter/properties/DataTypeDetectionTest.java (hunk 14 above)
Comment (ckunki) [reply src/test/java/com/exasol/adapter/properties/DataTypeDetectionTest.java:38]: 👍

File: src/test/java/com/exasol/adapter/properties/TableCountLimitTest.java (hunk 16 above)
Comment (ckunki) [reply src/test/java/com/exasol/adapter/properties/TableCountLimitTest.java:44]: 👍

File: src/main/java/com/exasol/adapter/jdbc/JDBCAdapter.java [hunk 20]
```diff
@@ -124,32 +116,49 @@ private SchemaMetadata getRemoteMetadata(final ExaMetadata metadata, final Refre
      *
      * @param properties           adapter properties
      * @param remoteTableAllowList allow list for remote tables
-     * @param exasolMetadata       exasol metadata
+     * @param exasolMetadata       ExaMetadata
      * @return schema metadata
      * @throws PropertyValidationException if properties are invalid
+     *
+     * @deprecated Should be removed as search on GitHub did not detect any call to this method by any other virtual
+     *             schema.
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/adapter/jdbc/JDBCAdapter.java:124]: Please do not use this method.

File: src/main/java/com/exasol/adapter/properties/ValidatorChain.java (hunk 10 above)
Comment (ckunki) [reply src/main/java/com/exasol/adapter/properties/ValidatorChain.java:11]: 👍

File: src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java (hunk 12 above)
Comment (ckunki) [reply src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java:150]: 👍

File: src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java [hunk 21]
```diff
@@ -177,27 +175,29 @@ protected String quoteIdentifierWithDoubleQuotes(final String identifier) {
...
+        this.propertyValidators.validate(this.properties);
     }
 
     /**
      * Validate that all given properties are supported by the dialect.
      *
      * @throws PropertyValidationException if validation fails
+     *
+     * @deprecated Should be removed as search on GitHub did not detect any call to this method by any other virtual
+     *             schema.
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/adapter/dialects/AbstractSqlDialect.java:187]: @deprecated Please do not use this method.

File: doc/changes/changes_10.2.0.md [hunk 22]
```diff
@@ -13,7 +13,7 @@ Additionally the property validation has been refactored which enabled to mark t
 * `AbstractSqlDialect.validateBooleanProperty()`
 * `AbstractSqlDialect.validateCastNumberToDecimalProperty()`
 
-Due to this potentially change this release increments the major version number of VSCJDBC from 10 to 11.
+Due to this potentially breaking change this release increments the major version number of VSCJDBC from 10 to 11.
```
Comment (kaklakariada) [inline doc/changes/changes_10.2.0.md:17]:
```suggestion
```
There should be no breaking change, and we increment to 10.2.0

**Review by kaklakariada — APPROVED** [review]

**Review by kaklakariada — CHANGES_REQUESTED** [review]
Please check the changelog

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**